modules/imap/functions.php

```diff
@@ -1673,3 +1673,722 @@ function is_imap_archive_folder($server_id, $user_config, $current_folder) {
+        } else {
+            return array('success' => false, 'error' => 'SMTP module required for APWG reporting. Please enable the SMTP module.');
+        }
+    }
+
+    $temp_file = create_spam_report_temp_file($sanitized_message, $user_config, $session, 'apwg_');
+    
+    $body = '';
+    $mime = new Hm_MIME_Msg($apwg_email, $subject, $body, $from_email, false, '', '', '', '', $from_email);
+    
+    $attachment = array(
+        'name' => 'phishing.eml',
+        'type' => 'message/rfc822',
+        'size' => strlen($sanitized_message),
+        'filename' => $temp_file
+    );
+    
+    $mime->add_attachments(array($attachment));
+
+    $mime_message = $mime->get_mime_msg();
+    
+    // Clean up extra blank lines
+    $mime_message = preg_replace('/\r\n\r\n+/', "\r\n\r\n", $mime_message);
+
+    $encoding_result = fix_spam_report_encoding($mime_message, 'APWG');
+    $mime_message = $encoding_result['mime_message'];
+    $mime_body = $encoding_result['mime_body'];
+    $boundary = $encoding_result['boundary'];
+
+    @unlink($temp_file);
+    
+    $headers = extract_spam_report_headers($mime_message, $boundary);
+    
+    $smtp_result = send_spam_report_via_smtp($from_email, $apwg_email, $subject, $mime_body, $boundary, $user_config, $session, 'APWG', true);
+    if ($smtp_result !== false) {
+        return $smtp_result;
+    }
+
+    return send_spam_report_via_mail($apwg_email, $subject, $mime_body, $headers, 'APWG');
+}}
+
+/**
+ * Sanitize message source for spam reporting
+ */
+if (!hm_exists('sanitize_message_for_spam_report')) {
+function sanitize_message_for_spam_report($message_source, $user_config) {
+    $user_emails = array();
+    $imap_servers = $user_config->get('imap_servers', array());
+    foreach ($imap_servers as $server) {
+        if (isset($server['user'])) {
+            $user_emails[] = strtolower($server['user']);
+        }
+    }
+
+    $parts = explode("\r\n\r\n", $message_source, 2);
+    $headers = isset($parts[0]) ? $parts[0] : '';
+    $body = isset($parts[1]) ? $parts[1] : '';
+
+    if (!empty($user_emails)) {
+        foreach ($user_emails as $email) {
+            // Remove email from various headers
+            $headers = preg_replace('/\b' . preg_quote($email, '/') . '\b/i', '[REDACTED]', $headers);
+        }
+    }
+
+    $sensitive_headers = array('X-Original-From', 'X-Forwarded-For', 'X-Real-IP');
+    foreach ($sensitive_headers as $header) {
+        $headers = preg_replace('/^' . preg_quote($header, '/') . ':.*$/mi', '', $headers);
+    }
+
+    // Clean blank lines
+    $headers = preg_replace('/\r\n\r\n+/', "\r\n\r\n", $headers);
+
+    return $headers . "\r\n\r\n" . $body;
+}
+}
+
+/**
+ * Extract IP address from email message headers
+ * @param string $message_source Full email message source
+ * @return string|false IP address (IPv4 or IPv6) or false if not found
+ */
+if (!hm_exists('extract_ip_from_message')) {
+function extract_ip_from_message($message_source) {
```

Comment:
imap module already has methods to return message headers. Please spare the low-level parsing of mime messages whenever possible.